rp2040-hal/examples/i2c_async_irq.rs

```diff
@@ -0,0 +1,164 @@
+//! # I²C Example
+//!
+//! This application demonstrates how to talk to I²C devices with an RP2040.
+//! in an Async environment.
+//!
+//! This example sets a `waker_setter` and is able to sleep when the task is pending on i2c
+//! updates..
+//!
+//! It may need to be adapted to your particular board layout and/or pin assignment.
+//!
+//! See the `Cargo.toml` file for Copyright and license details.
+
+#![no_std]
+#![no_main]
+
+use core::{cell::RefCell, task::Waker};
+
+use critical_section::Mutex;
+
+// Ensure we halt the program on panic (if we don't mention this crate it won't
+// be linked)
+use panic_halt as _;
+
+// Alias for our HAL crate
+use rp2040_hal as hal;
+
+// Some traits we need
+use hal::fugit::RateExtU32;
+
+// Import required types & traits.
+use eh_async_1_0_alpha::i2c::I2c;
+use hal::{
+    gpio::{FunctionI2C, Pin, PullUp},
+    pac::{self, interrupt},
+    Clock,
+};
+
+/// The linker will place this boot block at the start of our program image. We
+/// need this to help the ROM bootloader get our code up and running.
+/// Note: This boot block is not necessary when using a rp-hal based BSP
+/// as the BSPs already perform this step.
+#[link_section = ".boot2"]
+#[used]
+pub static BOOT2: [u8; 256] = rp2040_boot2::BOOT_LOADER_GENERIC_03H;
+
+/// External high-speed crystal on the Raspberry Pi Pico board is 12 MHz. Adjust
+/// if your board has a different frequency
+const XTAL_FREQ_HZ: u32 = 12_000_000u32;
+
+/// Mutex holding an optional waker for the i2c peripiheral.
+static SAVED_WAKER: Mutex<RefCell<Option<Waker>>> = Mutex::new(RefCell::new(None));
+
+/// Waker setter method updating the waker of the async task.
+pub fn i2c_on_pending(waker: Waker) {
+    critical_section::with(|cs| {
+        *SAVED_WAKER.borrow_ref_mut(cs) = Some(waker);
+    });
+}
+
+fn i2c_on_cancel() {
+    critical_section::with(|cs| {
+        *SAVED_WAKER.borrow_ref_mut(cs) = None;
```

Comment:
Is this really needed? What would happen if it wasn't done?
- the `Waker` inside `SAVED_WAKER` would stay around. That doesn't cost any additional resources (the memory usage of `SAVED_WAKER` doesn't change).
- it may happen that a later `I2C1_IRQ` interrupt calls `Waker::wake(waker)`. That may lead to a spurious wakeups, but that should be cheap - probably cheaper than taking the critical section in `i2c_on_cancel`

Are there any adverse effects of keeping the waker around that I'm missing?